File: src/profile/hud/mainmenu.cpp
#include "mainmenu.h"
#include "../profile_internal.h"
#include "../games/rne/tilebackground.h"
#include "../games/rne/mainmenu.h"
#include "../games/mo6tw/mainmenu.h"
#include "../../log.h"
#include "../../window.h"

namespace Impacto {
namespace Profile {
namespace MainMenu {

using namespace Impacto::MainMenu;

MainMenuType Type = MainMenuType::None;

Sprite MenuEntriesSprites[MenuEntriesNumMax];
Sprite MenuEntriesHSprites[MenuEntriesNumMax];

int MenuEntriesNum;
int MenuEntriesHNum;
float MenuEntriesX;
float MenuEntriesXOffset;
float MenuEntriesFirstY;
float MenuEntriesYPadding;

static void GetMemberSpriteArray(Sprite* arr, uint32_t count,
                                 char const* name) {
  EnsurePushMemberOfType(name, kArrayType);

  if (TopVal().Size() != count) {
    ImpLog(LL_Fatal, LC_Profile, "Expected to have %d sprites for %s\n", count,
           name);
    Window::Shutdown();
  }

  for (uint32_t i = 0; i < count; i++) {
    arr[i] = EnsureGetArrayElementSprite(i);
  }

  Pop();
}

void Configure() {
  if (TryPushMember("MainMenu")) {
    AssertIs(kObjectType);

    Type = MainMenuType::_from_integral_unchecked(EnsureGetMemberInt("Type"));

    if (Type == +MainMenuType::RNE) {
      Implementation = RNE::MainMenu::Configure();
    } else if (Type == +MainMenuType::MO6TW) {
      Implementation = MO6TW::MainMenu::Configure();
    }

    MenuEntriesNum = EnsureGetMemberInt("MenuEntriesNum");
    MenuEntriesHNum = EnsureGetMemberInt("MenuEntriesHNum");
    MenuEntriesX = EnsureGetMemberFloat("MenuEntriesX");
    MenuEntriesXOffset = EnsureGetMemberFloat("MenuEntriesXOffset");
    MenuEntriesFirstY = EnsureGetMemberFloat("MenuEntriesFirstY");
    MenuEntriesYPadding = EnsureGetMemberFloat("MenuEntriesYPadding");

    if (MenuEntriesNum > 0) {
      GetMemberSpriteArray(MenuEntriesSprites, MenuEntriesNum,
                           "MenuEntriesSprites");
    }
    if (MenuEntriesHNum > 0) {
      GetMemberSpriteArray(MenuEntriesHSprites, MenuEntriesHNum,
                           "MenuEntriesHighlightedSprites");
    }

    Pop();
  }
}

}  // namespace MainMenu
}  // namespace Profile
}  // namespace Impacto

File: profiles/mo6tw/hud/mainmenu.js
root.MainMenu = {
    Type: MainMenuType.MO6TW,
    MainMenuBackgroundSprite: "MainMenuBackground",
    MainMenuX: 993,
    MainMenuY: 114,
    MenuEntriesSprites: [],
    MenuEntriesHighlightedSprite: "MenuEntriesHSprite",
    MenuEntriesNum: 7,
    MenuEntriesHNum: 0,
    MenuEntriesX: 1001,
    MenuEntriesXOffset: 100,
    MenuEntriesFirstY: 126,
    MenuEntriesYPadding: 54,
    MenuEntriesTargetWidth: 254
};

for (var i = 0; i < 7; i++) {
    root.Sprites["MainMenuEntry" + i] = {
        Sheet: "Data",
        Bounds: {
            X: 584,
            Y: i * 32,
            Width: root.MainMenu.MenuEntriesTargetWidth,
            Height: 30
        }
    };
    root.MainMenu.MenuEntriesSprites.push("MainMenuEntry" + i);
}

root.Sprites["MenuEntriesHSprite"] = {
    Sheet: "Data",
    Bounds: { X: 65, Y: 1, Width: 94, Height: 30 }
};

root.Sprites["MainMenuBackground"] = {
    Sheet: "Data",
    Bounds: { X: 1765, Y: 1, Width: 282, Height: 380 }
};
